### PR TITLE
lib.fetchers: add hash-normalization helpers

### PR DESCRIPTION
**Pull request by nicoonoclaste**
## Description of changes

- [x] Add functions to `lib.fetchers`, intended to deduplicate the logic needed for #325892
  - `normalizeHash` converts an attrset containing one of `hash`, `sha256`, ... into one containing `outputHash{,Algo}` ;
  - `withNormalizedHash` wraps a function with `normalizeHash`, updating its signature :
    it automatically converts a fetcher which accepts `outputHash{, Algo}` into the current convention.
  - [x] Comply with current documentation conventions.
  - [x] Optimize until there is no perf. regression, ~~and even a 6% improvement as of 582439bfc2793729823e7d993149df170be01dff.~~
- [x] Wrap `fetchRepoProject` in `withNormalizedHash`, and update `amdvlk` to use the `hash` attribute as working example.
- [x] Modify `buildBazelPackage` to accept `fetchAttrs.hash` (using `normalizedHash`) and update a few packages as working examples.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] Tested compilation of all affected packages using `nixpkgs-review`.
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by Aleksanaa — COMMENTED** [review]
Looks reasonable; missing doc comments (or we don't write it for "internal use"?); Should evaluate performance impact after converting some major fetchers

File: lib/fetchers.nix [hunk 1]
```diff
@@ -14,4 +14,60 @@
...
+
+    All other attributes in the set remain as-is.
+  */
+  normalizeHashes = {
+    hashes ? [ "sha256" "sha512" ]
+  }: args:
+    with builtins; with lib;
+    let
+      hNames = [ "hash" ] ++ hashes;
+      hAttrs = genAttrs hNames (const {});
```
Comment (Aleksanaa) [inline lib/fetchers.nix:28]:
```suggestion
    hashTypes ? [ "sri" "sha256" "sha512" ]
  }: args:
    with builtins; with lib;
    let
      hAttrs = genAttrs hashTypes (const {});
```

Maybe better? (I changed "hash" to "sri")

File: lib/fetchers.nix [hunk 2]
```diff
@@ -14,4 +14,60 @@
...
+        else
+          head _h
+      ;
+    in
+      if args ? "outputHash" then
+        args
+      else
+        removeAttrs args hNames // {
+          outputHash = h.value;
+          outputHashAlgo = if h.name == "hash" then null else h.name;
```
Comment (Aleksanaa) [inline lib/fetchers.nix:46]: Is `"hash"` already having special handling here?

File: lib/fetchers.nix (hunk 1 above)
Comment (Aleksanaa) [reply lib/fetchers.nix:28]: Oh, I got it, it's compatible with current `"hash"` attribute.

File: lib/fetchers.nix [hunk 3]
```diff
@@ -14,4 +14,59 @@
...
+
+
+
+  /**
+    Wraps a function which accepts `outputHash{,Algo}` into one which accepts `hash` or `sha{256,512}`
+  */
+  withNormalizedHashes = {
+    hashTypes ? [ "sha256" ]
+  }: fetcher:
+    with builtins; with lib;
```
Comment (Aleksanaa) [inline lib/fetchers.nix:59]: Also this

**Comment by fricklerhandwerk** [comment]
@nbraud thanks for the ping. Looks pretty good, hard to get better with what we currently have.

**Comment by Aleksanaa** [comment]
> @Aleksanaa if this passes your review, I'll do one last no-diff rebase (as a treat) and this should be good to merge.

I'm sorry I can't give final review as I'm not really familiar with `lib` function guidelines. The only thing arise to my mind now is testing if `hash = ""` (assuming `lib.fakeHash`) and `sha256 = ""` (also assuming `lib.fakeHash` and would error out with SRI hash) still works as expected

**Review by mkg20001 — APPROVED** [review]

**Pull request merged**